### PR TITLE
[Enhancement] support using alias as function param in select item

### PR DESCRIPTION
**Pull request by kaijianding**
## Why I'm doing:
support using alias as function param, eg: `select trim(c1) as a, trim(c2) as b, concat(a,',', b) from t1 where a != 'x'`
## What I'm doing:
rewrite slotRef in function params to the real expression
eg: `select trim(c1) as a, trim(c2) as b, concat(trim(c1),',', trim(c2)) from t1 where trim(c1) != 'x'`

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by Seaven** [comment]
The question is not only where can't use alias(whether a function alias or a column alias), the sql execute order in sr is more like MySQL: `from->where->aggregator->having->select`, if WHERE clause can use alias, it's meanings the execute order will be: `from->select->where->aggregator->having`, will take some question:
* if WHERE support use aliases, the aliases can be used in aggregates/window/subquery? 
* when the aliaes and column names overlap, which one should be used in aggregates/window/subquery?
only rewrite where is confusing, I think we should to resolve these issuse before the modify

**Comment by kaijianding** [comment]
> The question is not only where can't use alias(whether a function alias or a column alias), the sql execute order in sr is more like MySQL: `from->where->aggregator->having->select`, if WHERE clause can use alias, it's meanings the execute order will be: `from->select->where->aggregator->having`, will take some question:
> 
> * if WHERE support use aliases, the aliases can be used in aggregates/window/subquery?
> * when the aliaes and column names overlap, which one should be used in aggregates/window/subquery?
>   only rewrite where is confusing, I think we should to resolve these issuse before the modify

Actually, the idea is very simple: replacing refferred alias with the actual expression before plan is generated.
So, the order how a query is executed is not changed.

Alias in `group by` and `having` is already handled by default, that is the reason why I just handle the select item and where clause.

`if WHERE support use aliases, the aliases can be used in aggregates/window/subquery?`
yes, the aliases can be used in aggregates/subquery, but not in window function this time. 
please check the test cases in `SelectUsingAliasTest`

`when the aliaes and column names overlap, which one should be used in aggregates/window/subquery?`
by default, the column name is used.
A session variable `enable_groupby_use_output_alias` is already defined to tell if alias should be used rather than column name in `group by`.
In this case, when `enable_groupby_use_output_alias` is set to `true`, this pr follows the same rule, the alias will be used rather than column name.

@Seaven

**Comment by Seaven** [comment]
> > The question is not only where can't use alias(whether a function alias or a column alias), the sql execute order in sr is more like MySQL: `from->where->aggregator->having->select`, if WHERE clause can use alias, it's meanings the execute order will be: `from->select->where->aggregator->having`, will take some question:
> > 
> > * if WHERE support use aliases, the aliases can be used in aggregates/window/subquery?
> > * when the aliaes and column names overlap, which one should be used in aggregates/window/subquery?
> >   only rewrite where is confusing, I think we should to resolve these issuse before the modify
> 
> Actually, the idea is very simple: replacing refferred alias with the actual expression before plan is generated. So, the order how a query is executed is kept unchanged.
> 
> Alias in `group by` and `having` is already handled by default, that is the reason why I just handle the select item and where clause.
> 
> `if WHERE support use aliases, the aliases can be used in aggregates/window/subquery?` yes, the aliases can be used in aggregates/window/subquery. please check the test cases in `SelectUsingAliasTest`
> 
> `when the aliaes and column names overlap, which one should be used in aggregates/window/subquery?` by default, the column name is used. A session variable `enable_groupby_use_output_alias` is already defined to tell if alias should be used rather than column name in `group by`. In this case, when `enable_groupby_use_output_alias` is set to `true`, this pr follows the same rule, the alias will be used rather than column name.
> 
> @Seaven

`enable_groupby_use_output_alias` will take aggregate to use alias column, I know it's like you want, their need adjusting the order  of SQL-clause execute, the current implementation is trickly because aggregate and select are execute adjacent to each other, but where-clause execute ealier than the aggregate, it's different from the aggregate, and test case doesn't answer the these questions.
e.g. 1
```
select trim(t1a) as t1c, concat(t1c,',') from test_all_type a where t1c = "a" and a.t1a in (select b.t1a from test_all_type b where b.t1c = a.t1c);
```
the `a.t1c` in subquery should be really `t1a` column? or `trim(t1a)` ?

e.g. 2
```
select trim(t1a) as t1c, concat(t1c, "") over (partition by t1c) as moving_averagem test_all_type where t1c = 'abc'
```
the `a.t1c` in window should be really `t1a` column? or `trim(t1a)` ,  window is execute after where, if where clause can be use `trim(t1a)`, why can't the window use `trim(t1a)` ?
so I think only rewrite where is confusing

**Comment by Seaven** [comment]
I suggest add some flag to control the execution order of the clauses, it's will automatically change the visibility between the clauses. this is you want? 
if we only want the where clause to use aliases from the select, I think this is difficult to achieve. It will make others can't understand how to executed SQL and will easily lead to misunderstandings of query results, especially when dealing with columns that have the same name

**Comment by Seaven** [comment]
@kaijianding you can connect me by wechat: Seaven7, we can discuss the issue

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java [hunk 1]
```diff
@@ -1216,4 +1219,107 @@ public static void checkJoinEqual(Expr expr) {
...
+            if (sourceScope.tryResolveField(slotRef).isPresent() &&
+                    !session.getSessionVariable().getEnableGroupbyUseOutputAlias()) {
+                return slotRef;
+            }
+            // Use short circuit to avoid duplicated resolving of same alias
+            if (resolvedAliases.containsKey(ref)) {
+                return resolvedAliases.get(ref);
+            }
+            if (resolvingAlias.contains(ref)) {
+                throw new SemanticException("Cyclic aliases: " + ref);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/QueryAnalyzer.java:1273]: slotRef has nodePosition info. Please add it into the error message to help client find the problem.

**Review by packy92 — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**